### PR TITLE
Copy Subscribers to Prevent Unwanted Recursion

### PR DESCRIPTION
**Pull request by RatanRSur**
Occasionally, a callback given to a subscriber will complete a future which then runs another task which adds another callback to the subsriber list of the same type. By the time the first callback is finished, the new callback (which presumably wants another "real" instance of the event to happen) could be next up in the list of the first's subscribers callback iteration. This takes a snapshot of the subsribers when the desired event happens so there's no risk of following down the rabbit hole.

I collected some rough data about the size of the subscriber lists to see the penalty for copying them and I found that by the time a full sync began or a fast sync had started downloading headers, the maximum subscribers any one Subscribers class had was 7.

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/SyncTargetManager.java [hunk 1]
```diff
@@ -113,7 +113,7 @@ protected SyncTargetManager(
   protected abstract CompletableFuture<Optional<EthPeer>> selectBestAvailableSyncTarget();
 
   private CompletableFuture<SyncTarget> waitForPeerAndThenSetSyncTarget() {
-    return waitForNewPeer().handle((r, t) -> r).thenCompose((r) -> selectNewSyncTarget());
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/SyncTargetManager.java:116]: So this is the main change?  the rest are all style adjustments?

File: util/src/main/java/org/hyperledger/besu/util/Subscribers.java [hunk 2]
```diff
@@ -101,17 +102,19 @@ public boolean unsubscribe(final long subscriberId) {
    * @param action the action to perform for each subscriber
    */
   public void forEach(final Consumer<T> action) {
-    subscribers
-        .values()
+    ImmutableSet.copyOf(
+            // we copy here to ensure that our callback, which may add another subscriber on this
```
Comment (shemnon) [inline util/src/main/java/org/hyperledger/besu/util/Subscribers.java:106]: Actually, this is the main change.  Can we move all style changes to a separate PR since they masked where the real change was?

File: util/src/main/java/org/hyperledger/besu/util/Subscribers.java [hunk 3]
```diff
@@ -101,17 +102,19 @@ public boolean unsubscribe(final long subscriberId) {
...
+            // event, won't trigger that new additional subscriber
+            subscribers.values())
         .forEach(
             subscriber -> {
               try {
                 action.accept(subscriber);
-              } catch (Throwable throwable) {
+              } catch (Exception e) {
```
Comment (rain-on) [inline util/src/main/java/org/hyperledger/besu/util/Subscribers.java:113]: changing this from throwable to exception _seems_ important - i.e. prior to this change, throwables would be eaten by the "suppress" - now they will always be raised (which maybe the intent, but wanted to confirm).

Comment (shemnon) [reply util/src/main/java/org/hyperledger/besu/util/Subscribers.java:113]: The only relevant throwable suppressed are `Error`s, which generally speaking are non-recoverable (`OutOfMemoryError, `ThreadDeath`, `LinkageError`).  When those happen the VM is going down hard.

Comment (RatanRSur) [reply util/src/main/java/org/hyperledger/besu/util/Subscribers.java:113]: @rain-on Are you alright with leaving it as is?

File: util/src/main/java/org/hyperledger/besu/util/Subscribers.java (hunk 2 above)
Comment (RatanRSur) [reply util/src/main/java/org/hyperledger/besu/util/Subscribers.java:106]: removed all the other changes

File: util/src/main/java/org/hyperledger/besu/util/Subscribers.java (hunk 3 above)
Comment (rain-on) [reply util/src/main/java/org/hyperledger/besu/util/Subscribers.java:113]: 👍  all good.
